File: app/src/main/java/de/datlag/hotdrop/utils/FileUtil.java
package de.datlag.hotdrop.utils;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.DialogInterface;
import android.net.Uri;
import android.util.Base64;
import android.webkit.MimeTypeMap;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.obsez.android.lib.filechooser.ChooserDialog;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.zeroturnaround.zip.ZipUtil;

import java.io.BufferedInputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileFilter;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URLConnection;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Objects;

import de.datlag.hotdrop.R;

public class FileUtil {

    public static final int MIN_LARGE_HEAP_SIZE = 11000000;
    private static final int DEFAULT_BUFFER_SIZE = 1024 * 4;

    public static void chooseFile(Activity activity, FileChooseCallback fileChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String path, File pathFile) {
                        fileChooseCallback.onChosen(path, pathFile);
                    }
                })
                // to handle the back key pressed or clicked outside the dialog:
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    public void onCancel(DialogInterface dialog) {
                        dialog.cancel();
                    }
                })
                .build()
                .show();
    }

    public static void chooseFile(Activity activity, FileFilter fileFilter, FileChooseCallback fileChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withFilter(fileFilter)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String path, File pathFile) {
                        fileChooseCallback.onChosen(path, pathFile);
                    }
                })
                // to handle the back key pressed or clicked outside the dialog:
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    public void onCancel(DialogInterface dialog) {
                        dialog.cancel();
                    }
                })
                .build()
                .show();
    }

    public static void chooseFolder(Activity activity, FolderChooseCallback folderChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withFilter(true, false)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String path, File pathFile) {
                        folderChooseCallback.onChosen(path, pathFile);
                    }
                })
                // to handle the back key pressed or clicked outside the dialog:
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    public void onCancel(DialogInterface dialog) {
                        dialog.cancel();
                    }
                })
                .build()
                .show();
    }

    public static void chooseFile(Activity activity, String startingPath, FileChooseCallback fileChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withStartFile(startingPath)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String path, File pathFile) {
                        fileChooseCallback.onChosen(path, pathFile);
                    }
                })
                // to handle the back key pressed or clicked outside the dialog:
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    public void onCancel(DialogInterface dialog) {
                        dialog.cancel();
                    }
                })
                .build()
                .show();
    }

    public static void chooseFolder(Activity activity, String startingPath, FolderChooseCallback folderChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withFilter(true, false)
                .withStartFile(startingPath)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String path, File pathFile) {
                        folderChooseCallback.onChosen(path, pathFile);
                    }
                })
                // to handle the back key pressed or clicked outside the dialog:
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    public void onCancel(DialogInterface dialog) {
                        dialog.cancel();
                    }
                })
                .build()
                .show();
    }

    public static void chooseAny(Activity activity, AnyChooseCallback anyChooseCallback) {
        new ChooserDialog(activity, R.style.FileChooserStyle)
                .withFilter(false, true)
                .withChosenListener(new ChooserDialog.Result() {
                    @Override
                    public void onChoosePath(String dir, File dirFile) {
                        if (isFile(dir)) {
                            anyChooseCallback.onChosenFile(dir, dirFile);
                        } else if (isDirectory(dir)) {
                            anyChooseCallback.onChosenFolder(dir, dirFile);
                        }
                    }
                })
                .withOnCancelListener(new DialogInterface.OnCancelListener() {
                    @Override
                    public void onCancel(DialogInterface dialogInterface) {
                        dialogInterface.cancel();
                    }
                })
                .withNegativeButton("Choose Folder", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        new ChooserDialog.Result() {
                            @Override
                            public void onChoosePath(String dir, File dirFile) {
                                if (isFile(dir)) {
                                    anyChooseCallback.onChosenFile(dir, dirFile);
                                } else if (isDirectory(dir)) {
                                    anyChooseCallback.onChosenFolder(dir, dirFile);
                                }
                            }
                        };
                    }
                })
                .build().show();
    }

    public static JsonObject jsonObjectFromFile(@NotNull Context context, File file, @NotNull ArrayList<byte[]> bytes, int pos) {
        String extension = MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(file).toString());
        JsonObject jsonObject = new JsonObject();
        ArrayList<String> base64List = new ArrayList<>();
        for (byte[] byteArray : bytes) {
            base64List.add(FileUtil.toBase64String(byteArray));
        }

        jsonObject.addProperty(context.getString(R.string.path), file.getAbsolutePath());
        jsonObject.addProperty(context.getString(R.string.name), file.getName());
        jsonObject.addProperty(context.getString(R.string.mime), MimeTypes.getMimeType(extension));
        jsonObject.addProperty(context.getString(R.string.extension), extension);
        jsonObject.addProperty(context.getString(R.string.base64), FileUtil.toBase64String(bytes.get(pos)));
        jsonObject.addProperty(context.getString(R.string.size), file.length());
        jsonObject.addProperty(context.getString(R.string.part), pos);
        jsonObject.addProperty(context.getString(R.string.maxParts), bytes.size());
        return jsonObject;
    }

    @NotNull
    public static byte[] jsonObjectToBytes(@NotNull JsonObject jsonObject) {
        return jsonObject.toString().getBytes();
    }

    private static ArrayList<byte[]> splitFileToBytes(File file){
        ArrayList<byte[]> returnValue = new ArrayList<>();
        int chunkSize = MIN_LARGE_HEAP_SIZE;
        BufferedInputStream in = null;
        try {
            in = new BufferedInputStream(new FileInputStream(file));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        long fileSize = file.length();

        int subfile;
        for (subfile = 0; subfile < fileSize / chunkSize; subfile++)
        {
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

            for (int currentByte = 0; currentByte < chunkSize; currentByte++)
            {
                try {
                    outputStream.write(in.read());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            try {
                outputStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }

            returnValue.add(outputStream.toByteArray());
        }

        if (fileSize != chunkSize * (subfile - 1))
        {
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

            int b = 0;
            while (true) {
                try {
                    if (!((b = in.read()) != -1)) break;
                } catch (IOException e) {
                    e.printStackTrace();
                }
                outputStream.write(b);
            }

            try {
                outputStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }

            returnValue.add(outputStream.toByteArray());
        }

        try {
            in.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return returnValue;
    }

    public static ArrayList<byte[]> byteArraysFromFile(File file) {
        ArrayList<byte[]> byteArrayList = new ArrayList<>();
        if (isParted(file)) {
            byteArrayList = splitFileToBytes(file);
        } else {
            byteArrayList.add(readFileToByteArray(file));
        }
        return byteArrayList;
    }

    @NotNull
    @Contract("_ -> new")
    public static FileInputStream openInputStream(@NotNull File file) throws IOException {
        if (file.exists()) {
            if (file.isDirectory()) {
                throw new IOException("File '" + file + "' exists but is a directory");
            }
            if (!file.canRead()) {
                throw new IOException("File '" + file + "' cannot be read");
            }
        } else {
            throw new FileNotFoundException("File '" + file + "' does not exist");
        }
        return new FileInputStream(file);
    }

    @NotNull
    public static byte[] readFileToByteArray(File file) {
        InputStream in = null;
        try {
            try {
                in = openInputStream(file);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return toByteArray(in);
        } finally {
            closeQuietly(in);
        }
    }

    @NotNull
    public static byte[] toByteArray(InputStream input) {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        copy(input, output);
        return output.toByteArray();
    }

    public static int copy(InputStream input, OutputStream output) {
        long count = copyLarge(input, output);
        if (count > Integer.MAX_VALUE) {
            return -1;
        }
        return (int) count;
    }

    public static long copyLarge(@NotNull InputStream input, OutputStream output) {
        byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        long count = 0;
        int n = 0;
        while (true) {
            try {
                if (!(-1 != (n = input.read(buffer)))) break;
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                output.write(buffer, 0, n);
            } catch (IOException e) {
                e.printStackTrace();
            }
            count += n;
        }
        return count;
    }

    public static void closeQuietly(InputStream input) {
        try {
            if (input != null) {
                input.close();
            }
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }

    @NotNull
    public static String readableFileSize(long size) {
        if(size <= 0){
            return "0";
        }
        final String[] units = new String[] { "B", "kB", "MB", "GB", "TB" };
        int digitGroups = (int) (Math.log10(size)/Math.log10(1024));
        return new DecimalFormat("#,##0.#").format(size/Math.pow(1024, digitGroups)) + " " + units[digitGroups];
    }

    public static boolean isParted(@NotNull File file) {
        if (file.length() > MIN_LARGE_HEAP_SIZE) {
            return true;
        }
        return false;
    }

    public static void writeBytesToFile(byte[] bytes, String file) {
        FileOutputStream out = null;
        try {
            out = new FileOutputStream(file);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        try {
            out.write(bytes);
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            out.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void createFile(String path, String filename) {
        makeDir(path);
        File file = new File(path+File.separator+filename);

        if (!file.exists()) {
            try {
                file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static String getMimeType(byte[] bytes) {
        InputStream is = new BufferedInputStream(new ByteArrayInputStream(bytes));
        String mimeType = null;
        try {
            mimeType = URLConnection.guessContentTypeFromStream(is);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return mimeType;
    }

    public static String getMimeType(Activity activity, @NotNull Uri uri) {
        String mimeType = null;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            Path path = Paths.get(uri.toString());
            try {
                mimeType = Files.probeContentType(path);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            if (Objects.equals(uri.getScheme(), ContentResolver.SCHEME_CONTENT)) {
                ContentResolver cr = activity.getContentResolver();
                mimeType = cr.getType(uri);
            } else {
                String fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString());
                mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension.toLowerCase());
            }
        }
        return mimeType;
    }

    public static String toBase64String(@NotNull byte[] bytes) {
        return Base64.encodeToString(bytes, Base64.DEFAULT);
    }

    public static byte[] base64ToBytes(String base64String) {
        return Base64.decode(base64String, Base64.DEFAULT);
    }

    public static JsonObject jsonObjectFromBytes(byte[] bytes) {
        return new Gson().fromJson(new String(bytes), JsonObject.class);
    }

    public interface FileChooseCallback {
        void onChosen(String path, File file);
    }

    public interface FolderChooseCallback {
        void onChosen(String path, File file) throws IOException;
    }

    public interface AnyChooseCallback {
        void onChosenFolder(String path, File file);

        void onChosenFile(String path, File file);
    }

    public static void makeDir(String path) {
        if(!existsFile(path)) {
            File file = new File(path);
            file.mkdirs();
        }
    }

    public static boolean existsFile(String path) {
        File file = new File(path);
        return file.exists();
    }

    public static boolean existsFile(@NotNull File file) {
        return file.exists();
    }

    public static boolean isDirectory(String path) {
        if (!existsFile(path)) {
            return false;
        }

        return new File(path).isDirectory();
    }

    public static boolean isFile(String path) {
        if (!existsFile(path)) {
            return false;
        }

        return new File(path).isFile();
    }

    public static boolean isDirectory(File file) {
        if (!existsFile(file)) {
            return false;
        }

        return file.isDirectory();
    }

    public static boolean isFile(File file) {
        if (!existsFile(file)) {
            return false;
        }

        return file.isFile();
    }

    @NotNull
    public static File folderToFile(@NotNull Context context, @NotNull File file) {
        File tempDir = context.getCacheDir();
        File tempFile = null;
        try {
            tempFile = File.createTempFile(file.getName(), ".zip", tempDir);
        } catch (IOException e) {
            e.printStackTrace();
        }
        ZipUtil.pack(file, tempFile);
        Objects.requireNonNull(tempFile).deleteOnExit();

        return tempFile;
    }

    @NotNull
    public static File folderToFile(@NotNull Context context, String path) {
        File file = new File(path);
        File tempDir = context.getCacheDir();
        File tempFile = null;
        try {
            tempFile = File.createTempFile(file.getName(), ".zip", tempDir);
        } catch (IOException e) {
            e.printStackTrace();
        }
        ZipUtil.pack(file, tempFile);
        Objects.requireNonNull(tempFile).deleteOnExit();

        return tempFile;
    }

    public static String getFileExtension(File file) {
        return MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(file).toString());
    }
}

File: app/src/main/java/de/datlag/hotdrop/util/FileUtil.kt
package de.datlag.hotdrop.util

import android.app.Activity
import android.content.ContentResolver
import android.content.Context
import android.content.DialogInterface
import android.net.Uri
import android.os.Build
import android.util.Base64
import android.webkit.MimeTypeMap
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.obsez.android.lib.filechooser.ChooserDialog
import de.datlag.hotdrop.R
import org.jetbrains.annotations.Contract
import org.zeroturnaround.zip.ZipUtil
import java.io.*
import java.net.URLConnection
import java.nio.file.Files
import java.nio.file.Paths
import java.text.DecimalFormat
import java.util.*
import kotlin.math.log10
import kotlin.math.pow

object FileUtil {
    private const val MIN_LARGE_HEAP_SIZE = 11000000
    private const val DEFAULT_BUFFER_SIZE = 1024 * 4

    fun chooseFile(activity: Activity?, fileFilter: FileFilter? = null, startingPath: String? = null, fileChooseCallback: FileChooseCallback) {
        ChooserDialog(activity, R.style.FileChooserStyle)
                .enableDpad(true)
                .withFilter(fileFilter)
                .withStartFile(startingPath)
                .withChosenListener { path: String?, file: File? -> fileChooseCallback.onChosen(path, file) }
                .withOnCancelListener { obj: DialogInterface -> obj.cancel() }
                .build()
                .show()
    }

    fun chooseFolder(activity: Activity?, startingPath: String? = null, folderChooseCallback: FolderChooseCallback) {
        ChooserDialog(activity, R.style.FileChooserStyle)
                .enableDpad(true)
                .withFilter(true, false)
                .withStartFile(startingPath)
                .withChosenListener { path: String?, pathFile: File? ->
                    try {
                        folderChooseCallback.onChosen(path, pathFile)
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
                .withOnCancelListener { obj: DialogInterface -> obj.cancel() }
                .build()
                .show()
    }

    fun chooseAny(activity: Activity?, anyChooseCallback: AnyChooseCallback) {
        ChooserDialog(activity, R.style.FileChooserStyle)
                .enableDpad(true)
                .withFilter(false, true)
                .withChosenListener { dir: String?, dirFile: File? ->
                    if (isFile(dir)) {
                        anyChooseCallback.onChosenFile(dir, dirFile)
                    } else if (isDirectory(dir)) {
                        anyChooseCallback.onChosenFolder(dir, dirFile)
                    }
                }
                .withOnCancelListener { obj: DialogInterface -> obj.cancel() }
                .withNegativeButton("Choose Folder") { _: DialogInterface?, _: Int ->
                    ChooserDialog.Result { dir, dirFile ->
                        if (isFile(dir)) {
                            anyChooseCallback.onChosenFile(dir, dirFile)
                        } else if (isDirectory(dir)) {
                            anyChooseCallback.onChosenFolder(dir, dirFile)
                        }
                    }
                }
                .build().show()
    }

    fun jsonObjectFromFile(context: Context, file: File, bytes: ArrayList<ByteArray>, pos: Int): JsonObject {
        val extension = MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(file).toString())
        val jsonObject = JsonObject()
        val base64List = ArrayList<String>()
        for (byteArray in bytes) {
            base64List.add(toBase64String(byteArray))
        }
        jsonObject.addProperty(context.getString(R.string.path), file.absolutePath)
        jsonObject.addProperty(context.getString(R.string.name), file.name)
        jsonObject.addProperty(context.getString(R.string.mime), MimeTypes.getMimeType(extension))
        jsonObject.addProperty(context.getString(R.string.extension), extension)
        jsonObject.addProperty(context.getString(R.string.base64), toBase64String(bytes[pos]))
        jsonObject.addProperty(context.getString(R.string.size), file.length())
        jsonObject.addProperty(context.getString(R.string.part), pos)
        jsonObject.addProperty(context.getString(R.string.maxParts), bytes.size)
        return jsonObject
    }

    fun jsonObjectToBytes(jsonObject: JsonObject): ByteArray {
        return jsonObject.toString().toByteArray()
    }

    private fun splitFileToBytes(file: File): ArrayList<ByteArray> {
        val returnValue = ArrayList<ByteArray>()
        val chunkSize = MIN_LARGE_HEAP_SIZE
        var `in`: BufferedInputStream? = null
        try {
            `in` = BufferedInputStream(FileInputStream(file))
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
        val fileSize = file.length()
        var subfile = 0
        while (subfile < fileSize / chunkSize) {
            val outputStream = ByteArrayOutputStream()
            for (currentByte in 0 until chunkSize) {
                try {
                    outputStream.write(`in`!!.read())
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            try {
                outputStream.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            returnValue.add(outputStream.toByteArray())
            subfile++
        }
        if (fileSize != chunkSize * (subfile - 1).toLong()) {
            val outputStream = ByteArrayOutputStream()
            var b = 0
            while (true) {
                try {
                    if (`in`!!.read().also { b = it } == -1) break
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                outputStream.write(b)
            }
            try {
                outputStream.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            returnValue.add(outputStream.toByteArray())
        }
        try {
            `in`!!.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return returnValue
    }

    fun byteArraysFromFile(file: File): ArrayList<ByteArray> {
        var byteArrayList = ArrayList<ByteArray>()
        if (isParted(file)) {
            byteArrayList = splitFileToBytes(file)
        } else {
            byteArrayList.add(readFileToByteArray(file))
        }
        return byteArrayList
    }

    @Throws(IOException::class)
    fun openInputStream(file: File): FileInputStream {
        if (file.exists()) {
            if (file.isDirectory) {
                throw IOException("File '$file' exists but is a directory")
            }
            if (!file.canRead()) {
                throw IOException("File '$file' cannot be read")
            }
        } else {
            throw FileNotFoundException("File '$file' does not exist")
        }
        return FileInputStream(file)
    }

    private fun readFileToByteArray(file: File): ByteArray {
        var `in`: InputStream? = null
        return try {
            try {
                `in` = openInputStream(file)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            toByteArray(`in`)
        } finally {
            closeQuietly(`in`)
        }
    }

    private fun toByteArray(input: InputStream?): ByteArray {
        val output = ByteArrayOutputStream()
        copy(input!!, output)
        return output.toByteArray()
    }

    private fun copy(input: InputStream, output: OutputStream): Int {
        val count = copyLarge(input, output)
        return if (count > Int.MAX_VALUE) {
            -1
        } else count.toInt()
    }

    private fun copyLarge(input: InputStream, output: OutputStream): Long {
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        var count: Long = 0
        var n = 0
        while (true) {
            try {
                if (-1 == input.read(buffer).also { n = it }) break
            } catch (e: IOException) {
                e.printStackTrace()
            }
            try {
                output.write(buffer, 0, n)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            count += n.toLong()
        }
        return count
    }

    private fun closeQuietly(input: InputStream?) {
        try {
            input?.close()
        } catch (ioe: IOException) {
            ioe.printStackTrace()
        }
    }

    fun readableFileSize(size: Long): String {
        if (size <= 0) {
            return "0"
        }
        val units = arrayOf("B", "kB", "MB", "GB", "TB")
        val digitGroups = (log10(size.toDouble()) / log10(1024.0)).toInt()
        return DecimalFormat("#,##0.#").format(size / 1024.0.pow(digitGroups.toDouble())) + " " + units[digitGroups]
    }

    private fun isParted(file: File): Boolean {
        return file.length() > MIN_LARGE_HEAP_SIZE
    }

    fun writeBytesToFile(bytes: ByteArray?, file: String?) {
        var out: FileOutputStream? = null
        try {
            out = FileOutputStream(file!!)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
        try {
            out!!.write(bytes!!)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        try {
            out!!.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun createFile(path: String, filename: String) {
        makeDir(path)
        val file = File(path + File.separator + filename)
        if (!file.exists()) {
            try {
                file.createNewFile()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun getMimeType(bytes: ByteArray?): String? {
        val `is`: InputStream = BufferedInputStream(ByteArrayInputStream(bytes))
        var mimeType: String? = null
        try {
            mimeType = URLConnection.guessContentTypeFromStream(`is`)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return mimeType
    }

    fun getMimeType(activity: Activity, uri: Uri): String? {
        var mimeType: String? = null
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val path = Paths.get(uri.toString())
            try {
                mimeType = Files.probeContentType(path)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        } else {
            mimeType = if (uri.scheme == ContentResolver.SCHEME_CONTENT) {
                val cr = activity.contentResolver
                cr.getType(uri)
            } else {
                val fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString())
                MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension.toLowerCase())
            }
        }
        return mimeType
    }

    private fun toBase64String(bytes: ByteArray): String {
        return Base64.encodeToString(bytes, Base64.DEFAULT)
    }

    fun base64ToBytes(base64String: String?): ByteArray {
        return Base64.decode(base64String, Base64.DEFAULT)
    }

    fun jsonObjectFromBytes(bytes: ByteArray?): JsonObject {
        return Gson().fromJson(String(bytes!!), JsonObject::class.java)
    }

    private fun makeDir(path: String?) {
        if (!existsFile(path)) {
            val file = File(path!!)
            file.mkdirs()
        }
    }

    private fun existsFile(path: String?): Boolean {
        val file = File(path!!)
        return file.exists()
    }

    private fun existsFile(file: File): Boolean {
        return file.exists()
    }

    private fun isDirectory(path: String?): Boolean {
        return if (!existsFile(path)) {
            false
        } else File(path!!).isDirectory
    }

    private fun isFile(path: String?): Boolean {
        return if (!existsFile(path)) {
            false
        } else File(path!!).isFile
    }

    fun isDirectory(file: File): Boolean {
        return if (!existsFile(file)) {
            false
        } else file.isDirectory
    }

    fun isFile(file: File): Boolean {
        return if (!existsFile(file)) {
            false
        } else file.isFile
    }

    fun folderToFile(context: Context, file: File): File {
        val tempDir = context.cacheDir
        var tempFile: File? = null
        try {
            tempFile = File.createTempFile(file.name, ".zip", tempDir)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        ZipUtil.pack(file, tempFile)
        tempFile!!.deleteOnExit()
        return tempFile
    }

    fun folderToFile(context: Context, path: String?): File {
        val file = File(path!!)
        val tempDir = context.cacheDir
        var tempFile: File? = null
        try {
            tempFile = File.createTempFile(file.name, ".zip", tempDir)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        ZipUtil.pack(file, tempFile)
        tempFile!!.deleteOnExit()
        return tempFile
    }

    fun getFileExtension(file: File?): String {
        return MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(file).toString())
    }

    interface FileChooseCallback {
        fun onChosen(path: String?, file: File?)
    }

    interface FolderChooseCallback {
        @Throws(IOException::class)
        fun onChosen(path: String?, file: File?)
    }

    interface AnyChooseCallback {
        fun onChosenFolder(path: String?, file: File?)
        fun onChosenFile(path: String?, file: File?)
    }
}